### DISCUSSION TITLE
Nesting Shortcodes

### DISCUSSION THREAD

**Discussion by tusamni** [11ty/eleventy#2039] [Uncategorized]
I'm hoping that I'm missing something here. I've been struggling with figuring out how to make this piece of code reusable. It's a simple featured posts section using a shortcode for the 11ty image plugin.

```
{% for featured in collections.featuredShoots %}
<a href="{{ featured.url }}" class="relative">
	{% image 
		src=featured.data.hero,
		imgClass="object-cover rounded-sm h-[300px] md:h-[500px]",
		showTitle=false,
		showDesc=false
	%}
	<div class="absolute px-2 py-1 text-sm font-medium text-black capitalize align-middle bg-white rounded-sm dark:text-gray-200 dark:bg-gray-700 top-4 right-4">
		{{ featured.data.category }}
	</div>
	<div class="relative block p-6 space-y-6 bg-white shadow-2xl bg-opacity-90 dark:bg-opacity-90 dark:bg-gray-700 lg:rounded-l-sm lg:rounded-br-sm lg:absolute lg:-bottom-20 lg:left-16">
		<h3 class="text-xl font-normal dark:text-gray-100">{{ featured.data.title }}</h3>
		<p class="text-lg font-light dark:text-gray-300">{{ featured.data.description }}</p>
		<div class="flex flex-row items-center w-full space-x-2 text-lg font-semibold group dark:text-gray-100">
			<p>View Photos</p>
			<div class="transition duration-300 transform group-hover:translate-x-2 arrow">
				&rarr;
			</div>
		</div>
	</div>
</a>
{% endfor %}
```

My initial thought was to just create another shortcode, but this didn't work. So I tried a Nunjucks includes, but that wouldn't render the image shortcode.

**Comment by cfjedimaster** [comment]
I'm having difficulty understanding what is going on - can you explain what isn't working? Are you saying when you take the code above, put it in a file that gets included, everything _but_ the image works?

**Reply by tusamni** [reply]
Geez, I'm all over the place. You're right. I posted the wrong code.

Here's what I was hoping would work:

Featured shortcode:
```
/**
 * Retuns the HTML for recent content
 *
 * @param {string} url The url to link
 * @param {string} image The url to hero image
 * @param {string} category Content category
 * @param {string} title Content title
 * @param {string} desc Content description
 */

module.exports = function (url, image, category, title, desc) {
	return `
		<a href="${url}" class="relative">
			{% image
				src=${image},
				imgClass="object-cover rounded-sm h-[300px] md:h-[500px]",
				showTitle=false,
				showDesc=false
			%}

			<div class="absolute px-2 py-1 text-sm font-medium text-black capitalize align-middle bg-white rounded-sm dark:text-gray-200 dark:bg-gray-700 top-4 right-4">
				${category}
			</div>

			<div class="relative block p-6 space-y-6 bg-white shadow-2xl bg-opacity-90 dark:bg-opacity-90 dark:bg-gray-700 lg:rounded-l-sm lg:rounded-br-sm lg:absolute lg:-bottom-20 lg:left-16">
				<h3 class="text-xl font-normal dark:text-gray-100">${title}</h3>
				<p class="text-lg font-light dark:text-gray-300">${desc}</p>
				<div class="flex flex-row items-center w-full space-x-2 text-lg font-semibold group dark:text-gray-100">
					<p>Keep Reading</p>
					<div class="transition duration-300 transform group-hover:translate-x-2 arrow">
						&rarr;
					</div>
				</div>
			</div>
		</a>
	`;
}
```

Calling the shortcode:
```
{% for shoot in collections.featuredShoots %}
	{% featured 
		shoot.url,
		shoot.data.hero,
		shoot.data.category,
		shoot.data.title,
		shoot.data.description
	%}
{% endfor %}
```

In the end, the image url works, but it doesn't render the shortcode, screenshot attached.
![Screenshot_1](https://user-images.githubusercontent.com/9271145/138161574-c61fefad-b8bb-4cbc-a8a8-a2ee9f3f3a75.png)

**Reply by tusamni** [reply]
I understand that JS literal is processing the {% image %} shortcode as a string. So I guess I'm just wondering, how do I include a shortcode inside of a shortcode? Is that even possible?

**Reply by cfjedimaster** [reply]
Ah I see what you mean. So filters can access other filters (https://www.11ty.dev/docs/filters/#access-existing-filters), but I don't see access for that in shortcodes. You can try running `eleventyConfig.getShortcode` or `getShortCode`, but as it's not documented, I'm assuming you can't.

**Reply by pdehaan** [reply]
I don't think I'm particularly happy with it, but here was my solution. I copied the initial shortcode and logic from the eleventy-img repo, and then my wrapper async shortcode had to use `${ await imageShortcode(...) }`, which I probably should have done outside the template string (in hindsight). But I could see this being a bit trickier if your shortcodes were defined in separate files and not all in the .eleventy.js config file.

```js
const Image = require("@11ty/eleventy-img");

async function imageShortcode(kwargs) {
  let metadata = await Image(kwargs.src, {
    widths: [300, 600],
    formats: ["avif", "jpeg"]
  });

  let imageAttributes = {
    alt: kwargs.alt || "",
    sizes: kwargs.sizes,
    loading: "lazy",
    decoding: "async",
  };

  // You bet we throw an error on missing alt in `imageAttributes` (alt="" works okay)
  return Image.generateHTML(metadata, imageAttributes);
}

module.exports = function(eleventyConfig) {
  eleventyConfig.addNunjucksAsyncShortcode("image", imageShortcode);
  eleventyConfig.addLiquidShortcode("image", imageShortcode);
  eleventyConfig.addJavaScriptFunction("image", imageShortcode);

  eleventyConfig.addNunjucksAsyncShortcode("featuredShoot", async function (featured={}) {
    const img = {
      src: featured.data.hero,
    };

    return `<a href="${ featured.url }" class="relative">
      ${ await imageShortcode(img) }
      <div class="absolute px-2 py-1 text-sm font-medium text-black capitalize align-middle bg-white rounded-sm dark:text-gray-200 dark:bg-gray-700 top-4 right-4">
        ${ featured.data.category }
      </div>
      <div class="relative block p-6 space-y-6 bg-white shadow-2xl bg-opacity-90 dark:bg-opacity-90 dark:bg-gray-700 lg:rounded-l-sm lg:rounded-br-sm lg:absolute lg:-bottom-20 lg:left-16">
        <h3 class="text-xl font-normal dark:text-gray-100">${ featured.data.title }</h3>
        <p class="text-lg font-light dark:text-gray-300">${ featured.data.description }</p>
        <div class="flex flex-row items-center w-full space-x-2 text-lg font-semibold group dark:text-gray-100">
          <p>View Photos</p>
          <div class="transition duration-300 transform group-hover:translate-x-2 arrow">
            &rarr;
          </div>
        </div>
      </div>
    </a>`;
  });

  return {
    dir: {
      input: "src",
      output: "www"
    }
  }
};
```

**Reply by cfjedimaster** [reply]
That's kinda bad ass.

**Reply by pdehaan** [reply]
Possibly this:

```js
    const img = {
      src: featured.data.hero,
    };

    const imgHtml = await eleventyConfig.nunjucksAsyncShortcodes.image(img);
    return `<a href="${ featured.url }" class="relative">
      ${ imgHtml }
      <div class="absolute px-2 py-1 text-sm font-medium text-black capitalize align-middle bg-white rounded-sm dark:text-gray-200 dark:bg-gray-700 top-4 right-4">
        ${ featured.data.category }
      </div>
      ...
    `;
```

Although not sure how much @zachleat would hate me for potentially abusing `eleventyConfig.nunjucksAsyncShortcodes` directly, although I didn't see an obvious getter function similar to `getFilter`: https://github.com/11ty/eleventy/blob/48c026922a1b59f33c9a3d512b10db03c711439a/src/UserConfig.js#L224-L231

One thing I was/am struggling with was passing of positional arguments vs using the nunjucks specific kwargs style like you were using; ref https://www.11ty.dev/docs/languages/nunjucks/#shortcode-named-argument-syntax-(nunjucks-only). Not sure which is better. One is definitely easier to read, but is Nunjucks specific, so not as portable.

**Comment by tusamni** [comment]
Thanks Peter,

An option for shortcodes like the getFilter would be nice.

The named paramaters aren't a big deal. I was only using them since it makes it a little easier to remember what I'm passing into the shortcode.

I'm going to give this a shot. If I can't make it work, then I'll probably just revert to using a CSS class for code reuse on the featured post and maintain the {% image %} shortcode for now.

**Reply by pdehaan** [reply]
A couple other random bits of things I learned today&hellip;

This did **NOT** seem to work (didn't output anything). I don't exactly know why, but I blame async shortcodes.

```njk
{% for featured in collections.featuredShoots %}
  <!-- #{{ loop.index }} -->
  {%- include "featured_shoot.njk" -%}
{% endfor %}
```

A bit of poking in the Nunjucks docs gave me an excuse to try [`asyncEach`](https://mozilla.github.io/nunjucks/templating.html#asynceach) which seems to work for my use case:

```njk
{% asyncEach featured in collections.featuredShoots %}
  <!-- #{{ loop.index }} -->
  {% include "featured_shoot.njk" %}
{% endeach %}
```

Where my ./src/_includes/featured_shoot.njk file looks like this:

```njk
<a href="{{ featured.url }}" class="relative">
  {% image 
    src=featured.data.hero,
    imgClass="object-cover rounded-sm h-[300px] md:h-[500px]",
    showTitle=false,
    showDesc=false
  %}
  <div class="absolute px-2 py-1 text-sm font-medium text-black capitalize align-middle bg-white rounded-sm dark:text-gray-200 dark:bg-gray-700 top-4 right-4">
    {{ featured.data.category }}
  </div>
  <div class="relative block p-6 space-y-6 bg-white shadow-2xl bg-opacity-90 dark:bg-opacity-90 dark:bg-gray-700 lg:rounded-l-sm lg:rounded-br-sm lg:absolute lg:-bottom-20 lg:left-16">
    <h3 class="text-xl font-normal dark:text-gray-100">{{ featured.data.title }}</h3>
    <p class="text-lg font-light dark:text-gray-300">{{ featured.data.description }}</p>
    <div class="flex flex-row items-center w-full space-x-2 text-lg font-semibold group dark:text-gray-100">
      <p>View Photos</p>
      <div class="transition duration-300 transform group-hover:translate-x-2 arrow">
        &rarr;
      </div>
    </div>
  </div>
</a>
```

I wasn't sure if you were thinking of reusing the featured shoot singular, or if your shortcode should include the loop as well, but I think your snippet above was for displaying a single image. :shrug:

**Reply by pdehaan** [reply]
> An option for shortcodes like the getFilter would be nice.

Filed as https://github.com/11ty/eleventy/issues/2041. Feel free to add feedback and/or vote.
